File: plugins/room_scan/ios/Classes/RoomScanPlugin.swift
import Flutter
import UIKit

/// Bridges the Dart `RoomScanChannel` to the native engines.
///
/// One method channel for commands, one event channel for progress / final
/// payloads, and one platform-view factory that mounts the active engine's
/// view.
final class RoomScanPlugin: NSObject, FlutterPlugin {
    static let methodChannelName = "taskfleet/room_scan"
    static let eventChannelName = "taskfleet/room_scan/events"
    static let viewTypeName = "taskfleet/room_scan_view"

    // Singleton glue. The view factory needs to talk to the same plugin
    // instance that owns the engines.
    private static var shared: RoomScanPlugin?

    private let eventSink = RoomScanEventSink()
    private var session: RoomScanSession?

    static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = RoomScanPlugin()
        shared = plugin

        let method = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(plugin, channel: method)

        let events = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        events.setStreamHandler(plugin.eventSink)

        let factory = RoomScanViewFactory(plugin: plugin)
        registrar.register(factory, withId: viewTypeName)
    }

    static func attach(view: RoomScanPlatformView) {
        shared?.session?.attach(view: view)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getCapabilities":
            result(Capabilities.detect().toMap())
        case "start":
            guard let args = call.arguments as? [String: Any],
                  let captureId = args["captureId"] as? String else {
                result(FlutterError(code: "bad_args", message: "captureId required", details: nil))
                return
            }
            let multiRoom = (args["multiRoom"] as? Bool) ?? false
            startSession(captureId: captureId, multiRoom: multiRoom)
            result(nil)
        case "finish":
            session?.finish()
            result(nil)
        case "finishRoom":
            session?.finishRoom()
            result(nil)
        case "finishAllRooms":
            session?.finishAllRooms()
            result(nil)
        case "stop":
            session?.stop()
            result(nil)
        case "resume":
            session?.resume()
            result(nil)
        case "undoLastTap":
            // Tap engines return the remaining count; engines without
            // discrete taps return -1 so the Dart side knows to no-op.
            let remaining = session?.undoLastTap() ?? -1
            result(remaining)
        case "cancel":
            session?.cancel()
            session = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startSession(captureId: String, multiRoom: Bool) {
        // If we're continuing a multi-room session (the user picked
        // "Scan room N+1" in the betweenRooms phase), keep the existing
        // engine so the AR world frame and the captured room list carry
        // through. The Dart side reuses the same captureId so we can
        // detect the chain.
        if let active = session, active.captureId == captureId {
            active.continueWithNextRoom()
            return
        }
        session?.cancel()
        let caps = Capabilities.detect()
        let engine: RoomScanEngine
        switch caps.engine {
        case .roomPlan:
            if #available(iOS 16.0, *) {
                engine = RoomPlanEngine(
                    captureId: captureId,
                    multiRoom: multiRoom && caps.supportsMultiRoom,
                    sink: eventSink
                )
            } else {
                engine = ArKitTapEngine(captureId: captureId, sink: eventSink)
            }
        case .arKitPlaneTap:
            engine = ArKitTapEngine(captureId: captureId, sink: eventSink)
        default:
            eventSink.send([
                "type": "error",
                "kind": "unsupportedDevice",
                "message": "No supported room scan engine for this device.",
            ])
            return
        }
        let next = RoomScanSession(engine: engine, sink: eventSink, captureId: captureId)
        session = next
        next.start()
    }
}

/// Forwards engine callbacks to the Dart EventChannel sink. Thread-safe in
/// the sense that all sends are marshalled to the main thread.
final class RoomScanEventSink: NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.sink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.sink = nil
        return nil
    }

    func send(_ payload: [String: Any]) {
        let sink = self.sink
        DispatchQueue.main.async {
            sink?(payload)
        }
    }
}

/// Active scan session — owns one engine and routes its lifecycle through
/// the platform view.
final class RoomScanSession {
    private let engine: RoomScanEngine
    private let sink: RoomScanEventSink
    private weak var view: RoomScanPlatformView?
    let captureId: String

    init(engine: RoomScanEngine, sink: RoomScanEventSink, captureId: String) {
        self.engine = engine
        self.sink = sink
        self.captureId = captureId
    }

    /// Multi-room: continue an existing engine into another room.
    /// Dispatches to the engine's `start` so it can re-arm its
    /// capture session without losing the prior captures.
    func continueWithNextRoom() {
        engine.start()
    }

    func attach(view: RoomScanPlatformView) {
        self.view = view
        if let host = view.containerView {
            engine.mount(into: host)
        }
    }

    func start() {
        engine.start()
    }

    func finish() {
        engine.finish()
    }

    func stop() {
        engine.stop()
    }

    func resume() {
        engine.resume()
    }

    /// Returns -1 if the active engine doesn't track discrete taps.
    func undoLastTap() -> Int {
        return engine.undoLastTap()
    }

    func finishRoom() {
        engine.finishRoom()
    }

    func finishAllRooms() {
        engine.finishAllRooms()
    }

    func cancel() {
        engine.cancel()
        sink.send(["type": "cancelled"])
    }
}

// MARK: - Platform view

final class RoomScanViewFactory: NSObject, FlutterPlatformViewFactory {
    private weak var plugin: RoomScanPlugin?

    init(plugin: RoomScanPlugin) {
        self.plugin = plugin
        super.init()
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let view = RoomScanPlatformView(frame: frame)
        RoomScanPlugin.attach(view: view)
        return view
    }
}

final class RoomScanPlatformView: NSObject, FlutterPlatformView {
    let containerView: UIView?

    init(frame: CGRect) {
        let v = UIView(frame: frame)
        v.backgroundColor = .black
        self.containerView = v
    }

    func view() -> UIView {
        return containerView ?? UIView()
    }
}

// MARK: - Capabilities

struct Capabilities {
    enum Engine: String {
        case roomPlan
        case arKitPlaneTap
        case none
    }

    let engine: Engine
    let supportsMultiRoom: Bool
    let hasDepthSensor: Bool
    let unsupportedReason: String?

    static func detect() -> Capabilities {
        #if canImport(RoomPlan)
        if #available(iOS 16.0, *) {
            let roomPlanSupported = RoomPlanCompat.isSupported()
            if roomPlanSupported {
                return Capabilities(
                    engine: .roomPlan,
                    supportsMultiRoom: roomPlanMultiRoomSupported(),
                    hasDepthSensor: true,
                    unsupportedReason: nil
                )
            }
        }
        #endif

        #if canImport(ARKit)
        if ArKitCompat.isWorldTrackingSupported() {
            return Capabilities(
                engine: .arKitPlaneTap,
                supportsMultiRoom: false,
                hasDepthSensor: false,
                unsupportedReason: nil
            )
        }
        #endif

        return Capabilities(
            engine: .none,
            supportsMultiRoom: false,
            hasDepthSensor: false,
            unsupportedReason: "AR is not supported on this device."
        )
    }

    func toMap() -> [String: Any] {
        var map: [String: Any] = [
            "engine": engine.rawValue,
            "supportsMultiRoom": supportsMultiRoom,
            "hasDepthSensor": hasDepthSensor,
        ]
        if let reason = unsupportedReason {
            map["unsupportedReason"] = reason
        }
        return map
    }

    private static func roomPlanMultiRoomSupported() -> Bool {
        if #available(iOS 17.0, *) {
            return true
        }
        return false
    }
}

File: plugins/room_scan/android/src/main/kotlin/com/taskfleet/room_scan/RoomScanPlugin.kt
package com.taskfleet.room_scan

import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import com.google.ar.core.ArCoreApk
import com.google.ar.core.Config
import com.google.ar.core.Session
import com.taskfleet.room_scan.engine.ArCorePlaneTapEngine
import com.taskfleet.room_scan.engine.RoomScanEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec

/**
 * Bridges the Dart `RoomScanChannel` to the native ARCore engine.
 *
 * Mirrors the iOS plugin's channel contract one-for-one:
 *   `taskfleet/room_scan`         — MethodChannel (commands)
 *   `taskfleet/room_scan/events`  — EventChannel  (progress + result)
 *   `taskfleet/room_scan_view`    — platform view factory
 */
class RoomScanPlugin :
    FlutterPlugin,
    ActivityAware,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    companion object {
        const val METHOD_CHANNEL = "taskfleet/room_scan"
        const val EVENT_CHANNEL = "taskfleet/room_scan/events"
        const val VIEW_TYPE = "taskfleet/room_scan_view"
    }

    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var sink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private var applicationContext: Context? = null
    private var activity: Activity? = null

    private var engine: RoomScanEngine? = null
    private var activeView: RoomScanPlatformView? = null
    private var captureId: String? = null

    // ── FlutterPlugin ──────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel!!.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel!!.setStreamHandler(this)

        binding.platformViewRegistry.registerViewFactory(
            VIEW_TYPE,
            RoomScanViewFactory(StandardMessageCodec.INSTANCE, this)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        applicationContext = null
        engine?.cancel()
        engine = null
    }

    // ── ActivityAware ──────────────────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        engine?.stop()
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        engine?.cancel()
        engine = null
        activity = null
    }

    // ── Platform view callback ─────────────────────────────────────────

    internal fun onViewCreated(view: RoomScanPlatformView) {
        activeView = view
        engine?.attach(view)
    }

    internal fun onViewDestroyed(view: RoomScanPlatformView) {
        if (activeView === view) {
            engine?.detach()
            activeView = null
        }
    }

    // ── MethodChannel ──────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getCapabilities" -> result.success(detectCapabilities())
            "start" -> {
                val id = call.argument<String>("captureId")
                if (id == null) {
                    result.error("bad_args", "captureId required", null)
                    return
                }
                startEngine(id)
                result.success(null)
            }
            "finish" -> {
                engine?.finish()
                result.success(null)
            }
            "stop" -> {
                engine?.stop()
                result.success(null)
            }
            "resume" -> {
                engine?.resume()
                result.success(null)
            }
            "undoLastTap" -> {
                // Tap engines return remaining count; engines without
                // discrete taps return -1 (Dart side no-ops on that).
                result.success(engine?.undoLastTap() ?: -1)
            }
            "cancel" -> {
                engine?.cancel()
                engine = null
                emit(mapOf("type" to "cancelled"))
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun startEngine(id: String) {
        val ctx = applicationContext ?: return
        val act = activity
        if (act == null) {
            emit(
                mapOf(
                    "type" to "error",
                    "kind" to "unknown",
                    "message" to "No activity available to start scan."
                )
            )
            return
        }
        captureId = id
        engine?.cancel()
        val newEngine = ArCorePlaneTapEngine(
            captureId = id,
            activity = act,
            emit = ::emit,
            depthRequested = isDepthSupported(ctx)
        )
        engine = newEngine
        activeView?.let { newEngine.attach(it) }
        newEngine.start()
    }

    private fun emit(payload: Map<String, Any?>) {
        val sink = this.sink ?: return
        mainHandler.post { sink.success(payload) }
    }

    // ── EventChannel ───────────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.sink = events
    }

    override fun onCancel(arguments: Any?) {
        this.sink = null
    }

    // ── Capabilities ───────────────────────────────────────────────────

    private fun detectCapabilities(): Map<String, Any?> {
        val ctx = applicationContext ?: return mapOf(
            "engine" to "none",
            "unsupportedReason" to "Plugin not attached."
        )
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            return mapOf(
                "engine" to "none",
                "unsupportedReason" to "Android 7.0 or later is required."
            )
        }
        val availability = try {
            ArCoreApk.getInstance().checkAvailability(ctx)
        } catch (t: Throwable) {
            return mapOf("engine" to "none", "unsupportedReason" to (t.message ?: "ARCore check failed"))
        }
        val supported = availability.isSupported || availability.isTransient
        if (!supported) {
            return mapOf(
                "engine" to "none",
                "unsupportedReason" to "ARCore is not available on this device."
            )
        }
        val depth = isDepthSupported(ctx)
        return mapOf(
            "engine" to if (depth) "arCoreDepth" else "arCorePlaneTap",
            "supportsMultiRoom" to false,
            "hasDepthSensor" to depth
        )
    }

    private fun isDepthSupported(ctx: Context): Boolean {
        return try {
            val session = Session(ctx)
            val supported = session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)
            session.close()
            supported
        } catch (_: Throwable) {
            false
        }
    }
}
